ambry-store/src/main/java/com/github/ambry/store/DiskManager.java

```diff
@@ -199,6 +202,33 @@ void start(boolean shouldRemoveUnexpectedDirs) throws InterruptedException {
       }
       if (numStoreFailures.get() > 0) {
         logger.error("Could not start {} out of {} stores on the disk {}", numStoreFailures.get(), stores.size(), disk);
+        if (storeConfig.storeRemoveDirectoryAndRestartBlobStore && numStoreFailures.get() != stores.size()) {
```

Comment:
What's the reasons so far the BlobStore cannot start?
If it's one log segment or index segment is broken, it's reasonable to remove it.

But is there any other possibilities which cause almost all BlobStore fails to boot up?
If there is any software bug which blocks booting, we'll remove all the blobs. It's not recoverable. 

I see that you have the check the numberStoreFailures != stores.size and there is shouldRemoveDirectory check.

Can we be more cautions? I don't have a good solution yet. 
1. Maybe we add one threshold, like  failedBlob threshold, only when the failed number is under that threshold, we'll remove blobs?
2. Or Just every bootup, no matter how many failed blobs, we only recover one.

Reply:
we currently have several error cases in clusters.

1. EndOffset of log is less than end offset of index
2. Recovery results in duplicate puts
3. Some index segment file only exist in directory metadata, but doesn't really exist.
4. File corruptions

These are the ones that I saw right now.

Reply:
Is that possible to pass these "error messages" into the Exception content? And then we only wipe out disk if it fails to these listed error messages.

Reply:
I will add some new error code and use these code to decide whether to remove the directory